### ISSUE TITLE
When context SDWebImageContextImageThumbnailPixelSize | SDWebImageContextImageTransformer is used, it will cause images to be downloaded repeatedly

### ISSUE THREAD

**Issue by helele90** [SDWebImage/SDWebImage#3217]
### New Issue Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | ios
 Platform Version        | e.g. 12.0 / 10.14.0 / 12.0 / 5.0
 SDWebImage Version      | e.g. 5.0.0 / 4.4.0
 Integration Method      | cocoapods
 Xcode Version           | Xcode 11
 Repro rate              | e.g. all the time (100%) / sometimes x% / only once
 Repro with our demo prj | e.g. does it happen with our demo project?
 Demo project link       | e.g. link to a demo project that highlights the issue

### Issue Description and Steps

If I set the `context` SDWebImageContextImageThumbnailPixelSize |  SDWebImageContextImageTransformer parameter, the cache key will contain related parameters. However, the disk cache does not require a new cache key, which will cause the images to be downloaded repeatedly when the parameters are different.
So the disk cache key should not need to use context to generate the key

example:
url: https://s1-relay.360buyimg.com/relay/web/img/img-wrap.6f18eb6d.png
context: SDWebImageContextImageThumbnailPixelSize CGSize(60, 60)-cacheKey: https://s1-relay.360buyimg.com/relay/web/img/img-wrap.6f18eb6d.png-Thumbnail(%7B60.000000,60.000000%7D,1).jpg 

context: SDWebImageContextImageThumbnailPixelSize CGSize(40, 40)-cacheKey: https://s1-relay.360buyimg.com/relay/web/img/img-wrap.6f18eb6d.png-Thumbnail(%7B40.000000,40.000000%7D,1).jpg 

``` objc
- (nullable NSString *)cacheKeyForURL:(nullable NSURL *)url context:(nullable SDWebImageContext *)context {
    if (!url) {
        return @"";
    }

    NSString *key;
    // Cache Key Filter
    id<SDWebImageCacheKeyFilter> cacheKeyFilter = self.cacheKeyFilter;
    if (context[SDWebImageContextCacheKeyFilter]) {
        cacheKeyFilter = context[SDWebImageContextCacheKeyFilter];
    }
    if (cacheKeyFilter) {
        key = [cacheKeyFilter cacheKeyForURL:url];
    } else {
        key = url.absoluteString;
    }

    // Thumbnail Key Appending
    NSValue *thumbnailSizeValue = context[SDWebImageContextImageThumbnailPixelSize];
    if (thumbnailSizeValue != nil) {
        CGSize thumbnailSize = CGSizeZero;
#if SD_MAC
        thumbnailSize = thumbnailSizeValue.sizeValue;
#else
        thumbnailSize = thumbnailSizeValue.CGSizeValue;
#endif
        BOOL preserveAspectRatio = YES;
        NSNumber *preserveAspectRatioValue = context[SDWebImageContextImagePreserveAspectRatio];
        if (preserveAspectRatioValue != nil) {
            preserveAspectRatio = preserveAspectRatioValue.boolValue;
        }
        key = SDThumbnailedKeyForKey(key, thumbnailSize, preserveAspectRatio);
    }

    // Transformer Key Appending
    id<SDImageTransformer> transformer = self.transformer;
    if (context[SDWebImageContextImageTransformer]) {
        transformer = context[SDWebImageContextImageTransformer];
        if (![transformer conformsToProtocol:@protocol(SDImageTransformer)]) {
            transformer = nil;
        }
    }
    if (transformer) {
        key = SDTransformedKeyForKey(key, transformer.transformerKey);
    }

    return key;
}

- (void)storeImage:(nullable UIImage *)image
         imageData:(nullable NSData *)imageData
            forKey:(nullable NSString *)key
          toMemory:(BOOL)toMemory
            toDisk:(BOOL)toDisk
        completion:(nullable SDWebImageNoParamsBlock)completionBlock {
    if (!image || !key) {
        if (completionBlock) {
            completionBlock();
        }
        return;
    }
    // if memory cache is enabled
    if (toMemory && self.config.shouldCacheImagesInMemory) {
        NSUInteger cost = image.jx_memoryCost;
        [self.memoryCache setObject:image forKey:key cost:cost];
    }

    if (!toDisk) {
        if (completionBlock) {
            completionBlock();
        }
        return;
    }
    dispatch_async(self.ioQueue, ^{
      @autoreleasepool {
          NSData *data = imageData;
          if (!data && [image conformsToProtocol:@protocol(SDAnimatedImage)]) {
              // If image is custom animated image class, prefer its original animated data
              data = [((id<SDAnimatedImage>)image) animatedImageData];
          }
          if (!data && image) {
              // Check image's associated image format, may return .undefined
              SDImageFormat format = image.jx_imageFormat;
              if (format == SDImageFormatUndefined) {
                  // If image is animated, use GIF (APNG may be better, but has bugs before macOS 10.14)
                  if (image.jx_isAnimated) {
                      format = SDImageFormatGIF;
                  } else {
                      // If we do not have any data to detect image format, check whether it contains alpha channel to use PNG or JPEG format
                      format = [SDImageCoderHelper CGImageContainsAlpha:image.CGImage] ? SDImageFormatPNG : SDImageFormatJPEG;
                  }
              }
              data = [[SDImageCodersManager sharedManager] encodedDataWithImage:image format:format options:nil];
          }
          [self _storeImageDataToDisk:data forKey:key];
          [self _archivedDataWithImage:image forKey:key];
      }

      if (completionBlock) {
          dispatch_async(dispatch_get_main_queue(), ^{
            completionBlock();
          });
      }
    });
}
```

**Comment by dreampiggy** [comment]
![image](https://user-images.githubusercontent.com/6919743/115507002-1da08380-a2ae-11eb-87aa-62804a427289.png)

Pass `SDWebImageContextOriginalStoreCacheType = .disk` as well, or we will not store the original image data into Disk Cache.

**Comment by dreampiggy** [comment]
Yes. We can make this `.originalStoreCache == .disk && .storeCacheType == .all` by default.

Currently we only set `.originalStoreCache == .none && .storeCacheType == .all`, which means:

+ Store the transformed image into memory cache; store the transformed image data, into disk cache
+ Does not store original image at all; Does not store original image data at all

**Comment by dreampiggy** [comment]
@helele90 See #3219.

You can actually change your API call with `storeCacheType = disk, queryCacheType = disk` by your own to use the same effect.

**Comment by helele90** [comment]
@dreampiggy thank you. let me try.
